articles/qos_deadline_liveliness_lifespan.md

```diff
@@ -0,0 +1,188 @@
+This may be acceptable for some QoS policies that only impact the behavior of the underlying DDS middleware, but in the case where support for QoS policy impacts the behavior of the application code this is not good enough.
+
+For example, it doesn't do much good to specify deadline policy via the vendor configuration mechanism if your ROS application is never notified when a publisher violates the deadline.
+In such cases the application needs to be able to respond to these events in order to log failures, notify users, or adjust system behavior.
+
+When interaction with the application is needed it is also preferable to have a ROS defined interface as opposed to having the application program directly against the DDS vendor API.
+Many nodes in ROS are written to be shared and reused by the community.
+Since different end users will want to use different middleware vendors for their applications, it is important to maintain that flexibility for nodes that want to take advantage of these additional QoS policies.
+
+**Based on this criteria, native ROS support should be provided for Deadline and Liveliness but not for the Lifespan policy.**
+While Deadline and Liveliness will both have events that the application will initiate and need to be informed of, the Lifespan policy can be entirely handled by the underlying DDS service without the need for any intervention by the application.
+
+### ROS Changes
+
+These are the various changes that would be needed within ROS in order to natively support Deadline and Liveliness.
+
+#### Topic Status Event Handler
+
+Both the Deadline and Liveliness policies generate events from the DDS service layer that the application will need to be informed of.
+For Deadlines, the subscriber receives event notifications if it doesn't receive anything within the deadline and the publisher receives event notifications if it doesn't publish anything within the deadline.
+For Liveliness, subscribers receive events when the publisher they're listening to is no longer alive. Both of these fall under a category of "Topic Status Events". 
+
+To handle these notifications, a new callback function can be provided by the user that will be called any time a `TopicStatusEvent` occurs for a particular topic.
+It will receive as a parameter a `TopicStatusEvent` enum value that specifies what type of event took place and a timestamp of when the event took place.
+This callback function would be provided by the user's application when it calls the create function for publishers and subscribers.
+The constructors and create functions will be overloaded to make this new handler optional.
+
+The choice to use a callback function as opposed to another notification mechanism is to remain consistent with the existing ROS interface and not add a new method of waiting on ROS data/events.
```

Comment:
A way to manually check for deadline status events is also necessary. Deadlines are usually related to real-time performance of the system, so having a method to check if a deadline has been broken in deterministic time is required.

Reply:
I would agree it could be useful, but I don't know if I'd agree it's necessary at this point. I think that a manual check could be added later as it may also require a new way of interfacing with the RMW layer. Do you really care about the realtime notification of deadline violations now when most the existing message notification method doesn't provide the same realtime determinism?

Reply:
Yes, I do care, just as I care about it for other message notifications. The lack of determinism in the existing message notification methods is a problem that needs to be addressed, but it's a separate issue. :smiley:

Reply:
I've added a new function to the design to work similar to how rmw_take works so that it can directly check for status events. I'll update the doc to also add a similar function analogous to rcl_take but probably won't go above that layer yet. 

That should set it up to better support direct checking in the future.

Reply:
It's a good start. We can take it further when we know more about how nodes will interact with these policies.

Reply:
@wjwwood are you talking about 2 methods here? 
1. wait as in waitset ?
2. poll as in https://community.rti.com/examples/polling-read ?